src/frontend/src/crypto/mnemonic.ts

```diff
@@ -1,18 +1,183 @@
+// NOTE: "english" is the only one actually bundled in (see webpack config)
+export const RECOVERYPHRASE_WORDLIST: string[] = wordlists.english;
+
 /**
  * @returns true if the mnemonic is valid, false otherwise.
  */
 export function validate(mnemonic: string): boolean {
   return validateMnemonic(mnemonic);
 }
+
+/** A recovery phrase that was parsed.
+ * An object of this type should not be used to reconstruct a recovery phrase, since
+ * some information (though most likely incorrectly inserted) like extra whitespaces
+ * is lost.
+ */
+type ParsedRecoveryPhrase = {
+  leadingUserNumber: string | null;
+  words: string[];
+};
+
+/** Parse recovery phrase into a structured object */
+export const parseRecoveryPhrase = (s: string): ParsedRecoveryPhrase => {
+  /* Split nicely into words
+   * ' 2002 foo  bar ' -> split(" ") -> [ '', '2002', 'foo', '', 'bar', '' ]
+   * [ '', '2002', 'foo', '', 'bar', '' ] -> filter(...) -> [ '2002', 'foo', 'bar' ]
+   */
+  const words = s.split(" ").filter((w) => w !== "");
+
+  // check if there's a leading user (anchor) number
+  const leadingUserNumber: string | null =
+    words.length >= 1 && isUserNumber(words[0]) ? words[0] : null;
+
+  // If the first element is indeed a number (BigInt), drop it from the actual mnemonic
+  if (leadingUserNumber !== null) {
+    words.shift();
+  }
+
+  return { leadingUserNumber, words };
+};
+
+/** Drop the anchor number from a string
+ * Example: " 10005 foo bar" -> "foo bar"
+ */
+export const dropLeadingUserNumber = (input: string): string => {
+  const parsed = parseRecoveryPhrase(input);
+  return parsed.leadingUserNumber !== null
+    ? input.replace(parsed.leadingUserNumber, "").trim()
+    : input;
+};
+
+/** The warning types that may be generated. For more information on each, see 'recoveryPhraseWarnings'.
+ *
+ * We generate the messages independently mostly so that we can programatically test the warnings generated
+ * (and avoid brittle regex-based testing).*/
+export type Warning =
+  | { type: "bad_chars"; chars: string[] }
+  | { type: "repeated_whitespace"; between: [string, string] | null }
+  | { type: "bad_anchor"; anchor: bigint }
+  | { type: "bad_word_count"; count: number }
+  | { type: "bad_words"; words: string[] }
+  | { type: "invalid" };
+
+/** Generates warnings for a given recovery phrase and user number */
+export const recoveryPhraseWarnings: {
+  (userNumber: bigint, inputRaw: string): Warning | null;
+}[] = [
+  // The input contains some characters that are not expected in a recovery phrase
+  (userNumber: bigint, inputRaw: string): Warning | null => {
+    // Check for anything that's not a letter, a number or a whitespace (' ', not \n)
+    const badChars = inputRaw.match(/[^a-z0-9\s]|[\n\t\r]/g);
+
+    if (badChars) {
+      return { type: "bad_chars", chars: [...new Set(badChars)] };
+    }
+
+    return null;
+  },
+
+  // Check for double (or triple) whitespaces between words
```

Comment:
Can we not correct for this by separately trimming each of the words?

Reply:
That's the first implementation I had, but I realized this would prevent users from submitting potentially correct recovery phrases, see comment [here](https://github.com/dfinity/internet-identity/pull/634/files#diff-f0a101a30b74057e65b16a3c5fb06d51da051c459c3efcd9a14be61a15aabdf2R31-R35) for instance:

> [we shouldn't use e.g. `words.join(" ")`] since some information (though most likely incorrectly inserted) like extra whitespaces is lost.